### PR TITLE
changefeed (ticdc): fix mysql sink config contains protocol error

### LINKED ISSUE #6874: ticdc v6.1.0 upgrade to v6.2.0 ,cdc changefeed status [CDC:ErrSinkURIInvalid]sink uri invalid 'protocol default is incompatible with mysql scheme'
### What did you do?

tiup ctl:v6.2.0 cdc changefeed list --pd=http://:2379/ 
Starting component `ctl`: /home/tidb/.tiup/components/ctl/v6.2.0/ctl cdc changefeed list --pd=http://:2379/
[
  {
    "id": "simple-replication-task",
    "namespace": "default",
    "summary": {
      "state": "stopped",
      "tso": 435502461638934530,
      "checkpoint": "2022-08-24 11:00:49.587",
      "error": {
        "addr": "[:8300](http://:8300/)",
        "code": "CDC:ErrSinkURIInvalid",
        "message": "[CDC:ErrSinkURIInvalid]sink uri invalid 'protocol default is incompatible with mysql scheme'"
      }
    }

cdc.toml 
case-sensitive = true
enable-old-value = true

[filter]
rules = ['*.*', '!test.*','!INFORMATION_SCHEMA.*','!METRICS_SCHEMA.*','!PERFORMANCE_SCHEMA.*','!mysql.*']

[mounter]
worker-num = 16

[sink]

dispatchers = [
    {matcher = ['*.*', '!test.*','!INFORMATION_SCHEMA.*','!METRICS_SCHEMA.*','!PERFORMANCE_SCHEMA.*','!mysql.*'], dispatcher = "ts"},
]
protocol = "default"

### What did you expect to see?

_No response_

### What did you see instead?

tiup ctl:v6.2.0 cdc  changefeed update --pd=http://:2379 --sink-uri="mysql://root:tidb@:34000/" --changefeed-id="simple-replication-task" --config=cdc.toml 

cdc.toml 
case-sensitive = true
enable-old-value = true

[filter]
rules = ['*.*', '!test.*','!INFORMATION_SCHEMA.*','!METRICS_SCHEMA.*','!PERFORMANCE_SCHEMA.*','!mysql.*']

[mounter]
worker-num = 16

[sink]

dispatchers = [
    {matcher = ['*.*', '!test.*','!INFORMATION_SCHEMA.*','!METRICS_SCHEMA.*','!PERFORMANCE_SCHEMA.*','!mysql.*'], dispatcher = "ts"},
]
protocol = ""

### Versions of the cluster

./cdc version
Release Version: v6.2.0
Git Commit Hash: c0eaa0075d46e3011225796c7f060e2a5d903a77
Git Branch: heads/refs/tags/v6.2.0
UTC Build Time: 2022-08-08 10:05:10
Go Version: go version go1.18.5 linux/amd64
Failpoint Build: false

### PR DESCRIPTION
**Pull request by asddongmen**
### What problem does this PR solve?

Issue Number: close #6874 

#### Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by CharlesCheung96 — APPROVED** [review]

File: cdc/model/changefeed.go [hunk 1]
```diff
@@ -292,9 +292,15 @@ func (info *ChangeFeedInfo) FixIncompatible() {
 
 	if creatorVersionGate.ChangefeedAcceptUnknownProtocols() {
 		log.Info("Start fixing incompatible changefeed sink protocol", zap.String("changefeed", info.String()))
```
Comment (hi-rustin) [inline cdc/model/changefeed.go:294]:
```suggestion
		log.Info("Start fixing incompatible changefeed MQ sink protocol", zap.String("changefeed", info.String()))
```

File: cdc/model/changefeed.go [hunk 2]
```diff
@@ -292,9 +292,15 @@ func (info *ChangeFeedInfo) FixIncompatible() {
...
-		info.fixSinkProtocol()
+		info.fixMQSinkProtocol()
 		log.Info("Fix incompatibility changefeed sink protocol completed", zap.String("changefeed", info.String()))
```
Comment (hi-rustin) [inline cdc/model/changefeed.go:296]:
```suggestion
		log.Info("Fix incompatibility changefeed MQ sink protocol completed", zap.String("changefeed", info.String()))
```

File: cdc/model/changefeed.go [hunk 3]
```diff
@@ -292,9 +292,15 @@ func (info *ChangeFeedInfo) FixIncompatible() {
...
 	}
+
+	if creatorVersionGate.ChangefeedAcceptProtocolInMysqlSinURI() {
+		log.Info("Start fixing incompatible changefeed sink uri", zap.String("changefeed", info.String()))
+		info.fixMysqlSinkProtocol()
```
Comment (hi-rustin) [inline cdc/model/changefeed.go:301]:
```suggestion
		info.fixMySQLSinkProtocol()
```

File: cdc/model/changefeed.go [hunk 4]
```diff
@@ -392,23 +392,38 @@ func (info *ChangeFeedInfo) fixSinkProtocol() {
...
+
+	if needsFix(info.Config.Sink.Protocol) {
+		log.Info("handle incompatible protocol from sink config",
+			zap.String("oldProtocol", info.Config.Sink.Protocol),
+			zap.String("fixedProtocol", openProtocol))
+		info.Config.Sink.Protocol = openProtocol
 	}
 }
 
+func (info *ChangeFeedInfo) fixSinkURIAndConfig(uri *url.URL, newProtocol string, newQuery url.Values) {
```
Comment (hi-rustin) [inline cdc/model/changefeed.go:414]:
```suggestion
func (info *ChangeFeedInfo) udapteSinkURIAndConfigProtcol(uri *url.URL, newProtocol string, newQuery url.Values) {
```

File: cdc/model/changefeed_test.go [hunk 5]
```diff
@@ -551,8 +550,62 @@ func TestFixSinkProtocol(t *testing.T) {
...
+		},
+	}
+
+	for _, tc := range sinkURITestCases {
+		tc.info.fixMysqlSinkProtocol()
+		require.Equal(t, tc.expectedSinkURI, tc.info.SinkURI)
+	}
+}
+
+func TestFixSinkProtocol(t *testing.T) {
```
Comment (hi-rustin) [inline cdc/model/changefeed_test.go:601]:
```suggestion
func TestFixMQSinkProtocol(t *testing.T) {
```

File: pkg/version/creator_version_gate.go [hunk 6]
```diff
@@ -67,13 +67,26 @@ func (f *CreatorVersionGate) ChangefeedStateFromAdminJob() bool {
...
+	if g.version == "" {
 		return true
 	}
 
-	creatorVersion := semver.New(removeVAndHash(f.version))
+	creatorVersion := semver.New(removeVAndHash(g.version))
 	return creatorVersion.LessThan(changefeedAcceptUnknownProtocolsVersion)
 }
+
+var changefeedAcceptProtocolInMysqlSinURI = *semver.New("6.2.0")
```
Comment (hi-rustin) [inline pkg/version/creator_version_gate.go:81]: What about `v6.1.1`?

**Review by nongfushanquan — APPROVED** [review]

File: cdc/model/changefeed.go [hunk 7]
```diff
@@ -349,39 +352,36 @@ func (info *ChangeFeedInfo) fixSinkProtocol() {
...
-			log.Warn("sink URI or sink config contains protocol, but scheme is not mq",
-				zap.String("sinkURI", maskedSinkURI),
-				zap.String("protocol", protocolStr),
-				zap.Any("sinkConfig", info.Config.Sink))
-			// always set protocol of mysql sink to ""
-			rawQuery.Del(config.ProtocolKey)
-			fixSinkURI("")
-		}
+
+	if sink.IsMQScheme(uri.Scheme) {
```
Comment (hi-rustin) [inline cdc/model/changefeed.go:356]: What happens if we got an s3 sink? Maybe we can improve it later.

Comment (asddongmen) [reply cdc/model/changefeed.go:356]: yep

**Pull request merged**